sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakeFileSystemAsyncClient.java

```diff
@@ -714,6 +805,86 @@ public HttpPipeline getHttpPipeline() {
+        return restorePathWithResponse(deletedPath, deletionId).flatMap(FluxUtil::toMono);
+    }
+
+    /**
+     * Restores a soft deleted path in the file system. For more information see the
+     * <a href="https://docs.microsoft.com/rest/api/storageservices/datalakestoragegen2/path/delete">Azure
+     * Docs</a>.
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * {@codesnippet com.azure.storage.file.datalake.DataLakeFileSystemAsyncClient.restorePathWithResponse#String-String}
+     *
+     * @param deletedPath The deleted path
+     * @param deletionId deletion ID associated with the soft deleted path that uniquely identifies a resource if
+     * multiple have been soft deleted at this location.
+     * You can get soft deleted paths and their associated deletion IDs with {@link #listDeletedPaths()}.
+     * @return A reactive response signalling completion.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Response<DataLakePathAsyncClient>> restorePathWithResponse(String deletedPath, String deletionId) {
+        try {
+            return withContext(context -> restorePathWithResponse(deletedPath, deletionId, context));
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    Mono<Response<DataLakePathAsyncClient>> restorePathWithResponse(String deletedPath, String deletionId,
+        Context context) {
+        Objects.requireNonNull(deletedPath);
+        Objects.requireNonNull(deletionId);
+
+        context = context == null ? Context.NONE : context;
+        String blobUrl = DataLakeImplUtils.endpointToDesiredEndpoint(blobDataLakeStorageFs.getUrl(), "blob", "dfs");
+        blobUrl = StorageImplUtils.appendToUrlPath(blobUrl, deletedPath).toString();
+        DataLakeStorageClientImpl blobDataLakeStoragePath = new DataLakeStorageClientBuilder()
+            .pipeline(blobDataLakeStorageFs.getHttpPipeline())
+            .url(blobUrl)
+            .version(serviceVersion.getVersion())
+            .build();
+
+        // Initial rest call
+        return blobDataLakeStoragePath.paths().undeleteWithRestResponseAsync(null,
+            String.format("?%s=%s", Constants.UrlConstants.DELETIONID_QUERY_PARAMETER, deletionId), null,
+            context.addData(AZ_TRACING_NAMESPACE_KEY, STORAGE_TRACING_NAMESPACE_VALUE))
+            /*
+            Perform a get properties to determine if it's a file or directory. Zip with the original response for
+            constructing the final response later
+             */
+            .flatMap(response -> Mono.zip(Mono.just(response),
+                this.blobContainerAsyncClient.getBlobAsyncClient(deletedPath, null).getProperties())
+            .onErrorMap(DataLakeImplUtils::transformBlobStorageException)
+            // Construct the new client and final response from the undelete + getProperties responses
+            .map(tuple2 -> new SimpleResponse<>(tuple2.getT1(),
```

Comment:
This should be able to just use the return header on the api rather than make a get properties call.